libsrc/osic1p/cgetc.c

```diff
@@ -0,0 +1,167 @@
+
+    for (row = 0; row <= 7; row++) { // Scan each row.
+
+        writeRow(row); // Select row
+        data = ~PEEK(KEYBOARD); // Read column data for row.
+
+        // Handle modifier keys on row 0.
+        if (row == 0) {
+            if (data & controlMask) { // Control key pressed.
+                control = 1;
+            }
+            if (data & shiftMask) { // Left Shift or Right Shift pressed.
+                shift = 1;
+            }
+            if (data & lockMask) { // Shift Lock pressed.
+                lock = 1;
+            }
+        }
+
+        data &= columnMask[row]; // Mask out only the valid bits for the row.
+ 
+        if (data != 0) {
+            // Find bit for first pressed key.
+            for (bit = 7; bit >= 0; bit--) {
+                if (data & (1 << bit)) {
+                    break;
+                }
+            }
+
+            key = keys[row][bit]; // Get key character.
+
+            if (key == 0) { // Check for invalid key.
+                continue; // Shouldn't be possible, but just in case (user may have modified hardware).
+            }
+
+            // Handle possible control key. Takes precedence over shift.
+            if (control) {
+                key &= 0x1f; // Convert to control code.
+            } else {
+                // Handle possible caps lock key.
+                if (lock) {
+                    key = toupper(key); // Convert to upper case.
+                }
+                // Handle possible shift key.
+                if (shift) {
+                    if (key >= 'a' && key <= 'z') {
+                        key -= 32; // Lower to upper case.
+                    } else if (key >= '1' && key <= ';') {
+                        key -= 16; // Number keys and punctuation.
+                    } else if (key >= ',' && key <= '/') {
+                        key += 16; // Punctuation.
+                    } else if (key >= 'K' && key <= 'O') {
+                        key += 16; // KLMNO become [\]^_
+                    } else if (key == 'P') {
+                        key = '@'; // P becomes @
+                    }
+                }
+            }
+
+            // Return the key.
+            return key;
+        }
+    }
+
+    return 0; // No key pressed.
+}
+
+/*
+ * Return true (1) if key pressed, otherwise false (0). Saves
+ * character so that it is guaranteed that if kbhit() returns true, a
+ * subsequent call to cgetc() will not block.
+ */
+unsigned char kbhit (void)
+{
+    lastKey = getkey();    // Get possible key press.
+    return (lastKey != 0); // Return whether key was pressed.
+}
+
+/*
+ * Return key. Blocks until key pressed. Uses key from a previous call
+ * to kbhit(), if any.
+ */
+char cgetc (void)
+{
+    unsigned char key;
+    
+    if (lastKey != 0) { // Did we previously get a key?
+        key = lastKey; // If so, use it.
+        lastKey = 0; // Then clear it.
+    } else {
+        while ((key = getkey()) != 0)
```

Comment:
The loop must be:

    while ((key = getkey()) == 0)
        ; // Wait for key to be pressed.

Otherwise cgetc() will not block.